File: gp/src/main/java/uk/healthtechwales/gppractice/service/DocumentProcessingService.java
package uk.healthtechwales.gppractice.service;

import org.hl7.fhir.r4.model.Binary;
import org.hl7.fhir.r4.model.Bundle;
import org.hl7.fhir.r4.model.Identifier;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import uk.healthtechwales.gppractice.config.ProcessingMode;
import uk.healthtechwales.gppractice.model.QuarantineMessage;

@Service
public class DocumentProcessingService {

    private static final Logger log = LoggerFactory.getLogger(DocumentProcessingService.class);

    private final ProcessingMode processingMode;
    private final ClamAVService clamAVService;
    private final MinIOService minioService;
    private final RabbitMQService rabbitMQService;

    public DocumentProcessingService(
            @Value("${document.processing.mode}") String mode,
            ClamAVService clamAVService,
            MinIOService minioService,
            RabbitMQService rabbitMQService) {
        this.processingMode = ProcessingMode.valueOf(mode.toUpperCase());
        this.clamAVService = clamAVService;
        this.minioService = minioService;
        this.rabbitMQService = rabbitMQService;
        
        log.info("Document processing mode: {}", this.processingMode);
    }

    /**
     * Process document based on configured mode
     * @return ProcessingResult
     */
    public ProcessingResult processDocument(
            String gpPracticeId,
            Bundle bundle,
            String bundleJson,
            Binary binary,
            String base64Data) {
        
        switch (processingMode) {
            case SYNC:
                return processSynchronous(base64Data);
                
            case ASYNC:
                return processAsynchronous(gpPracticeId, bundle, bundleJson, binary, base64Data);
                
            case HYBRID:
                return processHybrid(gpPracticeId, bundle, bundleJson, binary, base64Data);
                
            default:
                throw new IllegalStateException("Unknown processing mode: " + processingMode);
        }
    }

    /**
     * SYNC Mode: Check binary with ClamAV, respond based on result
     */
    private ProcessingResult processSynchronous(String base64Data) {
        log.info("Processing in SYNC mode");
        
        ClamAVService.ScanResult scanResult = clamAVService.scanDocument(base64Data);
        
        if (!scanResult.isClean()) {
            return ProcessingResult.builder()
                    .success(false)
                    .errorMessage("Malware detected in submitted document: " + scanResult.getDetails())
                    .malwareDetected(true)
                    .build();
        }
        
        return ProcessingResult.builder()
                .success(true)
                .scanStatus("CLEAN")
                .build();
    }

    /**
     * ASYNC Mode: Validate syntax/format, respond immediately, 
     * store binary in MinIO and send to RabbitMQ for offline processing
     */
    private ProcessingResult processAsynchronous(
            String gpPracticeId,
            Bundle bundle,
            String bundleJson,
            Binary binary,
            String base64Data) {
        
        log.info("Processing in ASYNC mode");
        
        // Extract patient information
        String nhsNumber = extractNHSNumber(bundle);
        String documentReferenceId = extractDocumentReferenceId(bundle);
        
        // Store in MinIO
        MinIOService.StorageResult storageResult = minioService.storeDocument(
                base64Data, 
                nhsNumber, 
                binary.getContentType());
        
        if (!storageResult.isSuccess()) {
            return ProcessingResult.builder()
                    .success(false)
                    .errorMessage("Failed to store document: " + storageResult.getErrorMessage())
                    .build();
        }
        
        // Send to RabbitMQ
        try {
            QuarantineMessage message = QuarantineMessage.builder()
                    .objectName(storageResult.getObjectName())
                    .bucketName(storageResult.getBucketName())
                    .etag(null)
                    .size(storageResult.getSize())
                    .contentType(storageResult.getContentType())
                    .submissionId(java.util.UUID.randomUUID().toString())
                    .patientId(nhsNumber)
                    .documentReferenceId(documentReferenceId)
                    .originalFilename(storageResult.getObjectName())
                    .build();
            
            rabbitMQService.sendQuarantineMessage(message);
            
            return ProcessingResult.builder()
                    .success(true)
                    .scanStatus("PENDING_ASYNC")
                    .storageResult(storageResult)
                    .build();
                    
        } catch (Exception e) {
            log.error("Failed to send message to RabbitMQ", e);
            return ProcessingResult.builder()
                    .success(false)
                    .errorMessage("Failed to queue document for processing: " + e.getMessage())
                    .build();
        }
    }

    /**
     * HYBRID Mode: Check binary with ClamAV, respond based on result,
     * if clean store in MinIO and send to RabbitMQ for deeper inspection
     */
    private ProcessingResult processHybrid(
            String gpPracticeId,
            Bundle bundle,
            String bundleJson,
            Binary binary,
            String base64Data) {
        
        log.info("Processing in HYBRID mode");
        
        // First, scan with ClamAV
        ClamAVService.ScanResult scanResult = clamAVService.scanDocument(base64Data);
        
        if (!scanResult.isClean()) {
            return ProcessingResult.builder()
                    .success(false)
                    .errorMessage("Malware detected in submitted document: " + scanResult.getDetails())
                    .malwareDetected(true)
                    .build();
        }
        
        // Document is clean, now store and queue for deeper inspection
        String nhsNumber = extractNHSNumber(bundle);
        String documentReferenceId = extractDocumentReferenceId(bundle);
        
        // Store in MinIO
        MinIOService.StorageResult storageResult = minioService.storeDocument(
                base64Data, 
                nhsNumber, 
                binary.getContentType());
        
        if (!storageResult.isSuccess()) {
            return ProcessingResult.builder()
                    .success(false)
                    .errorMessage("Failed to store document: " + storageResult.getErrorMessage())
                    .build();
        }
        
        // Send to RabbitMQ for deeper inspection
        try {
            QuarantineMessage message = QuarantineMessage.builder()
                    .objectName(storageResult.getObjectName())
                    .bucketName(storageResult.getBucketName())
                    .etag(null)
                    .size(storageResult.getSize())
                    .contentType(storageResult.getContentType())
                    .submissionId(java.util.UUID.randomUUID().toString())
                    .patientId(nhsNumber)
                    .documentReferenceId(documentReferenceId)
                    .originalFilename(storageResult.getObjectName())
                    .build();
            
            rabbitMQService.sendQuarantineMessage(message);
            
            return ProcessingResult.builder()
                    .success(true)
                    .scanStatus("CLEAN")
                    .storageResult(storageResult)
                    .build();
                    
        } catch (Exception e) {
            log.error("Failed to send message to RabbitMQ", e);
            // Document is still clean and accepted, but couldn't queue for deep scan
            log.warn("Document accepted but not queued for deep inspection");
            return ProcessingResult.builder()
                    .success(true)
                    .scanStatus("CLEAN")
                    .storageResult(storageResult)
                    .build();
        }
    }

    private String extractNHSNumber(Bundle bundle) {
        return bundle.getEntry().stream()
                .filter(entry -> entry.getResource() instanceof org.hl7.fhir.r4.model.Patient)
                .map(entry -> (org.hl7.fhir.r4.model.Patient) entry.getResource())
                .flatMap(patient -> patient.getIdentifier().stream())
                .filter(identifier -> identifier.hasSystem() && 
                        identifier.getSystem().equals("https://fhir.nhs.uk/Id/nhs-number"))
                .map(Identifier::getValue)
                .findFirst()
                .orElse("UNKNOWN");
    }

    private String extractDocumentReferenceId(Bundle bundle) {
        return bundle.getEntry().stream()
                .filter(entry -> entry.getResource() instanceof org.hl7.fhir.r4.model.DocumentReference)
                .map(entry -> (org.hl7.fhir.r4.model.DocumentReference) entry.getResource())
                .filter(docRef -> docRef.hasId())
                .map(docRef -> docRef.getId())
                .findFirst()
                .orElse(java.util.UUID.randomUUID().toString());
    }

    public static class ProcessingResult {
        private boolean success;
        private String scanStatus;
        private String errorMessage;
        private boolean malwareDetected;
        private MinIOService.StorageResult storageResult;

        public ProcessingResult() {}

        private ProcessingResult(Builder builder) {
            this.success = builder.success;
            this.scanStatus = builder.scanStatus;
            this.errorMessage = builder.errorMessage;
            this.malwareDetected = builder.malwareDetected;
            this.storageResult = builder.storageResult;
        }

        public boolean isSuccess() {
            return success;
        }

        public String getScanStatus() {
            return scanStatus;
        }

        public String getErrorMessage() {
            return errorMessage;
        }

        public boolean isMalwareDetected() {
            return malwareDetected;
        }

        public MinIOService.StorageResult getStorageResult() {
            return storageResult;
        }

        public static Builder builder() {
            return new Builder();
        }

        public static class Builder {
            private boolean success;
            private String scanStatus;
            private String errorMessage;
            private boolean malwareDetected;
            private MinIOService.StorageResult storageResult;

            public Builder success(boolean success) {
                this.success = success;
                return this;
            }

            public Builder scanStatus(String scanStatus) {
                this.scanStatus = scanStatus;
                return this;
            }

            public Builder errorMessage(String errorMessage) {
                this.errorMessage = errorMessage;
                return this;
            }

            public Builder malwareDetected(boolean malwareDetected) {
                this.malwareDetected = malwareDetected;
                return this;
            }

            public Builder storageResult(MinIOService.StorageResult storageResult) {
                this.storageResult = storageResult;
                return this;
            }

            public ProcessingResult build() {
                return new ProcessingResult(this);
            }
        }
    }
}

File: src/gp_practice_documents/services/document_processing_service.py
import logging
import uuid
from typing import Optional

from fhir.resources.binary import Binary
from fhir.resources.bundle import Bundle
from fhir.resources.documentreference import DocumentReference
from fhir.resources.patient import Patient
from pydantic import BaseModel

from ..config import ProcessingMode, Settings
from ..models import QuarantineMessage
from .clamav_service import ClamAVService
from .minio_service import MinIOService, StorageResult
from .rabbitmq_service import RabbitMQService

logger = logging.getLogger(__name__)


class ProcessingResult(BaseModel):
    """Result of document processing."""
    success: bool
    scan_status: Optional[str] = None
    error_message: Optional[str] = None
    malware_detected: bool = False
    storage_result: Optional[StorageResult] = None


class DocumentProcessingService:
    """Service for processing documents based on configured mode."""

    def __init__(
        self,
        settings: Settings,
        clamav_service: ClamAVService,
        minio_service: MinIOService,
        rabbitmq_service: RabbitMQService
    ):
        self.processing_mode = settings.processing_mode
        self.clamav_service = clamav_service
        self.minio_service = minio_service
        self.rabbitmq_service = rabbitmq_service

        logger.info(f"Document processing mode: {self.processing_mode}")

    def process_document(
        self,
        gp_practice_id: str,
        bundle: Bundle,
        bundle_json: str,
        binary: Binary,
        base64_data: str
    ) -> ProcessingResult:
        """
        Process document based on configured mode.

        Args:
            gp_practice_id: GP Practice ID
            bundle: FHIR Bundle
            bundle_json: Bundle as JSON string
            binary: Binary resource
            base64_data: Base64-encoded document data

        Returns:
            ProcessingResult
        """
        if self.processing_mode == ProcessingMode.SYNC:
            return self._process_synchronous(base64_data)
        elif self.processing_mode == ProcessingMode.ASYNC:
            return self._process_asynchronous(gp_practice_id, bundle, bundle_json, binary, base64_data)
        elif self.processing_mode == ProcessingMode.HYBRID:
            return self._process_hybrid(gp_practice_id, bundle, bundle_json, binary, base64_data)
        else:
            raise ValueError(f"Unknown processing mode: {self.processing_mode}")

    def _process_synchronous(self, base64_data: str) -> ProcessingResult:
        """SYNC Mode: Check binary with ClamAV, respond based on result."""
        logger.info("Processing in SYNC mode")

        scan_result = self.clamav_service.scan_document(base64_data)

        if not scan_result.is_clean():
            return ProcessingResult(
                success=False,
                error_message=f"Malware detected in submitted document: {scan_result.get_details()}",
                malware_detected=True
            )

        return ProcessingResult(
            success=True,
            scan_status="CLEAN"
        )

    def _process_asynchronous(
        self,
        gp_practice_id: str,
        bundle: Bundle,
        bundle_json: str,
        binary: Binary,
        base64_data: str
    ) -> ProcessingResult:
        """
        ASYNC Mode: Validate syntax/format, respond immediately,
        store binary in MinIO and send to RabbitMQ for offline processing.
        """
        logger.info("Processing in ASYNC mode")

        # Extract patient information
        nhs_number = self._extract_nhs_number(bundle)
        document_reference_id = self._extract_document_reference_id(bundle)

        # Store in MinIO
        storage_result = self.minio_service.store_document(
            base64_data,
            nhs_number,
            binary.contentType
        )

        if not storage_result.success:
            return ProcessingResult(
                success=False,
                error_message=f"Failed to store document: {storage_result.error_message}"
            )

        # Send to RabbitMQ
        try:
            message = QuarantineMessage(
                object_name=storage_result.object_name,
                bucket_name=storage_result.bucket_name,
                etag=None,
                size=storage_result.size,
                content_type=storage_result.content_type,
                submission_id=str(uuid.uuid4()),
                patient_id=nhs_number,
                document_reference_id=document_reference_id,
                original_filename=storage_result.object_name
            )

            self.rabbitmq_service.send_quarantine_message(message)

            return ProcessingResult(
                success=True,
                scan_status="PENDING_ASYNC",
                storage_result=storage_result
            )

        except Exception as e:
            logger.error(f"Failed to send message to RabbitMQ: {e}")
            return ProcessingResult(
                success=False,
                error_message=f"Failed to queue document for processing: {str(e)}"
            )

    def _process_hybrid(
        self,
        gp_practice_id: str,
        bundle: Bundle,
        bundle_json: str,
        binary: Binary,
        base64_data: str
    ) -> ProcessingResult:
        """
        HYBRID Mode: Check binary with ClamAV, respond based on result,
        if clean store in MinIO and send to RabbitMQ for deeper inspection.
        """
        logger.info("Processing in HYBRID mode")

        # First, scan with ClamAV
        scan_result = self.clamav_service.scan_document(base64_data)

        if not scan_result.is_clean():
            return ProcessingResult(
                success=False,
                error_message=f"Malware detected in submitted document: {scan_result.get_details()}",
                malware_detected=True
            )

        # Document is clean, now store and queue for deeper inspection
        nhs_number = self._extract_nhs_number(bundle)
        document_reference_id = self._extract_document_reference_id(bundle)

        # Store in MinIO
        storage_result = self.minio_service.store_document(
            base64_data,
            nhs_number,
            binary.contentType
        )

        if not storage_result.success:
            return ProcessingResult(
                success=False,
                error_message=f"Failed to store document: {storage_result.error_message}"
            )

        # Send to RabbitMQ for deeper inspection
        try:
            message = QuarantineMessage(
                object_name=storage_result.object_name,
                bucket_name=storage_result.bucket_name,
                etag=None,
                size=storage_result.size,
                content_type=storage_result.content_type,
                submission_id=str(uuid.uuid4()),
                patient_id=nhs_number,
                document_reference_id=document_reference_id,
                original_filename=storage_result.object_name
            )

            self.rabbitmq_service.send_quarantine_message(message)

            return ProcessingResult(
                success=True,
                scan_status="CLEAN",
                storage_result=storage_result
            )

        except Exception as e:
            logger.error(f"Failed to send message to RabbitMQ: {e}")
            # Document is still clean and accepted, but couldn't queue for deep scan
            logger.warning("Document accepted but not queued for deep inspection")
            return ProcessingResult(
                success=True,
                scan_status="CLEAN",
                storage_result=storage_result
            )

    @staticmethod
    def _extract_nhs_number(bundle: Bundle) -> str:
        """Extract NHS number from bundle."""
        for entry in bundle.entry:
            if isinstance(entry.resource, Patient):
                patient = entry.resource
                for identifier in patient.identifier or []:
                    if identifier.system == "https://fhir.nhs.uk/Id/nhs-number" and identifier.value:
                        return identifier.value
        return "UNKNOWN"

    @staticmethod
    def _extract_document_reference_id(bundle: Bundle) -> str:
        """Extract DocumentReference ID from bundle."""
        for entry in bundle.entry:
            if isinstance(entry.resource, DocumentReference):
                doc_ref = entry.resource
                if doc_ref.id:
                    return doc_ref.id
        return str(uuid.uuid4())
